File: src/versioning/FileBackend.js
const fs = require('fs').promises;
const path = require('path');
const execa = require('execa');
const VersioningBackend = require('./VersioningBackend');

class FileBackend extends VersioningBackend {
  constructor(execa) {
    super();
    this.files = new Map();
    this.execa = execa;
  }

  async add(file) {
    try {
      await fs.access(file);
      if (!this.files.has(file)) {
        const backupFile = `${file}.${Date.now()}.bak`;
        await fs.copyFile(file, backupFile);
        this.files.set(file, backupFile);
      }
    } catch (error) {
      throw new Error(`File not found: ${file}`);
    }
  }

  async unstage(file) {
    try {
      const backupFile = this.files.get(file);
      if (backupFile) {
        await fs.unlink(backupFile);
        this.files.delete(file);
      }
    } catch (error) {
      console.error(`Error unstaging file ${file}:`, error);
    }
  }

  async record(message) {
    // No-op for file backend
  }

  async unrecord(hash) {
    // No-op for file backend
  }

  async channel(name) {
    // No-op for file backend
  }

  async patch(subcommand, name) {
    // No-op for file backend
  }

  async apply(patch) {
    // No-op for file backend
  }

  async conflicts() {
    return '[]';
  }

  async revert(file) {
    try {
      const backupFile = this.files.get(file);
      if (backupFile) {
        await fs.copyFile(backupFile, file);
      }
    } catch (error) {
      console.error(`Error reverting file ${file}:`, error);
    }
  }

  async undo() {
    for (const file of this.files.keys()) {
      await this.revert(file);
    }
  }

  async diff() {
    let diff = '';
    for (const [file, backupFile] of this.files) {
      try {
        const { stdout } = await this.execa('diff', ['-u', backupFile, file], { reject: false });
        diff += stdout;
      } catch (error) {
        console.error(`Error getting diff for file ${file}:`, error);
      }
    }
    return diff;
  }

  async clear() {
    for (const backupFile of this.files.values()) {
      try {
        await fs.unlink(backupFile);
      } catch (error) {
        console.error(`Error clearing backups:`, error);
      }
    }
    this.files.clear();
  }
}

module.exports = FileBackend;

File: java/src/main/java/com/pijul/aider/versioning/FileBackend.java
package com.pijul.aider.versioning;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

public class FileBackend extends VersioningBackend {
    private final Map<String, String> files = new HashMap<>();

    @Override
    public CompletableFuture<Void> add(String file) {
        return CompletableFuture.runAsync(() -> {
            try {
                Path filePath = Paths.get(file);
                if (Files.exists(filePath)) {
                    if (!files.containsKey(file)) {
                        String backupFile = file + "." + System.currentTimeMillis() + ".bak";
                        Files.copy(filePath, Paths.get(backupFile), StandardCopyOption.REPLACE_EXISTING);
                        files.put(file, backupFile);
                    }
                } else {
                    throw new RuntimeException("File not found: " + file);
                }
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        });
    }

    @Override
    public CompletableFuture<Void> unstage(String file) {
        return CompletableFuture.runAsync(() -> {
            String backupFile = files.get(file);
            if (backupFile != null) {
                try {
                    Files.deleteIfExists(Paths.get(backupFile));
                    files.remove(file);
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }
            }
        });
    }

    @Override
    public CompletableFuture<Void> record(String message) {
        // No-op for file backend
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletableFuture<Void> unrecord(String hash) {
        // No-op for file backend
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletableFuture<Void> channel(String subcommand, String name) {
        // No-op for file backend
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletableFuture<Void> patch(String subcommand, String name) {
        // No-op for file backend
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletableFuture<Void> apply(String patch) {
    public CompletableFuture<Void> apply(String patch) {
        return CompletableFuture.runAsync(() -> {
            try {
                // Create temporary file for patch
                Path tempPatch = Files.createTempFile("patch", ".diff");
                Files.write(tempPatch, patch.getBytes(StandardCharsets.UTF_8));
                
                // Apply patch using external tool (e.g., git)
                ProcessBuilder processBuilder = new ProcessBuilder("git", "apply", tempPatch.toString());
                processBuilder.redirectErrorStream(true);
                Process process = processBuilder.start();
                
                // Wait for process to complete
                int exitCode = process.waitFor();
                if (exitCode != 0) {
                    throw new RuntimeException("Failed to apply patch: " + new String(process.getErrorStream().readAllBytes()));
                }
                
                // Clean up temporary file
                Files.delete(tempPatch);
            } catch (IOException | InterruptedException e) {
                throw new RuntimeException("Error applying patch", e);
            }
        });
    }
    }

    @Override
    public CompletableFuture<String> conflicts() {
        return CompletableFuture.completedFuture("[]");
    }

    @Override
    public CompletableFuture<Void> revert(String file) {
        return CompletableFuture.runAsync(() -> {
            String backupFile = files.get(file);
            if (backupFile != null) {
                try {
                    Files.copy(Paths.get(backupFile), Paths.get(file), StandardCopyOption.REPLACE_EXISTING);
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }
            }
        });
    }

    @Override
    public CompletableFuture<Void> undo() {
        return CompletableFuture.runAsync(() -> {
            for (String file : files.keySet()) {
                try {
                    revert(file).get();
                } catch (Exception e) {
                    throw new RuntimeException(e);
                }
            }
        });
    }

    @Override
    public CompletableFuture<String> diff() {
        // Simplified implementation - would need to implement diff logic
        return CompletableFuture.completedFuture("");
    }

    @Override
    public CompletableFuture<Void> clear() {
        return CompletableFuture.runAsync(() -> {
            for (String backupFile : files.values()) {
                try {
                    Files.deleteIfExists(Paths.get(backupFile));
                } catch (Exception e) {
                    // Ignore
                }
            }
            files.clear();
        });
    }

    @Override
    public CompletableFuture<List<String>> listTrackedFiles() {
        return CompletableFuture.completedFuture(new ArrayList<>(files.keySet()));
    }

    @Override
    public CompletableFuture<List<String>> listUntrackedFiles() {
        return CompletableFuture.completedFuture(new ArrayList<>());
    }
}